File: frontend/src/utils/ieltsBandScale.js
export const IELTS_MAX_BAND = 9;

const toNumber = (value) => {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : null;
};

export const usesBandScale = (assessment) => {
  const skill = String(assessment?.skill || '').toUpperCase();
  const type = String(assessment?.type || '').toUpperCase();
  const mode = String(assessment?.aiEvaluationMode || '').toUpperCase();

  if (skill === 'LISTENING' || skill === 'READING') {
    return false;
  }
  if (mode === 'ESTIMATED_BAND') {
    return true;
  }
  if (type !== 'MODULE_TEST' && type !== 'MOCK_TEST') {
    return false;
  }
  return true;
};

export const resolveScoreCap = (assessment) => {
  const configured = toNumber(assessment?.maxScore);
  if (!usesBandScale(assessment)) {
    return configured;
  }
  if (configured == null || configured > IELTS_MAX_BAND) {
    return IELTS_MAX_BAND;
  }
  return configured;
};

export const clampBand = (value) => {
  const parsed = toNumber(value);
  if (parsed == null) {
    return null;
  }
  const bounded = Math.max(0, Math.min(IELTS_MAX_BAND, parsed));
  return Math.round(bounded * 2) / 2;
};

export const normalizeAssessmentMaxScore = (assessment) => {
  const raw = toNumber(assessment?.maxScore);
  if (raw == null) {
    return usesBandScale(assessment) ? IELTS_MAX_BAND : null;
  }
  if (usesBandScale(assessment) && raw > IELTS_MAX_BAND) {
    return IELTS_MAX_BAND;
  }
  return raw;
};

export const normalizeBandThreshold = (assessment, raw) => {
  const parsed = toNumber(raw);
  if (parsed == null) {
    return null;
  }
  return usesBandScale(assessment) ? clampBand(parsed) : parsed;
};

export const normalizeAssessmentPassingScore = (assessment) => {
  const raw = toNumber(assessment?.passingScore);
  if (raw == null) {
    return null;
  }
  return usesBandScale(assessment) ? clampBand(raw) : raw;
};

File: backend/src/main/java/fu/sep490/g23/backend/service/assessment/IeltsBandScale.java
package fu.sap490.g23.backend.service.assessment;

import fu.sap490.g23.backend.entity.assessment.CourseAssessment;
import fu.sap490.g23.backend.entity.assessment.enums.AiEvaluationMode;
import fu.sap490.g23.backend.entity.assessment.enums.AssessmentSkill;
import fu.sap490.g23.backend.entity.assessment.enums.AssessmentType;

import java.math.BigDecimal;
import java.math.RoundingMode;

public final class IeltsBandScale {

    public static final BigDecimal MAX = BigDecimal.valueOf(9.0);

    private IeltsBandScale() {
    }

    public static boolean usesBandScale(CourseAssessment assessment) {
        if (assessment == null) {
            return false;
        }
        return usesBandScale(
                assessment.getType(),
                assessment.getSkill(),
                assessment.getAiEvaluationMode()
        );
    }

    public static boolean usesBandScale(
            AssessmentType type,
            AssessmentSkill skill,
            AiEvaluationMode evaluationMode
    ) {
        if (skill == AssessmentSkill.LISTENING || skill == AssessmentSkill.READING) {
            return false;
        }
        if (evaluationMode == AiEvaluationMode.ESTIMATED_BAND) {
            return true;
        }
        if (type != AssessmentType.MODULE_TEST && type != AssessmentType.MOCK_TEST) {
            return false;
        }
        return true;
    }

    public static BigDecimal resolveScoreCap(CourseAssessment assessment) {
        if (assessment == null) {
            return null;
        }
        if (!usesBandScale(assessment)) {
            return assessment.getMaxScore();
        }
        BigDecimal configured = assessment.getMaxScore();
        if (configured == null || configured.compareTo(MAX) > 0) {
            return MAX;
        }
        return configured;
    }

    public static BigDecimal normalizeBand(BigDecimal value) {
        if (value == null) {
            return null;
        }
        double bounded = Math.max(0, Math.min(MAX.doubleValue(), value.doubleValue()));
        return BigDecimal.valueOf(Math.round(bounded * 2) / 2.0).setScale(1, RoundingMode.HALF_UP);
    }

    public static BigDecimal normalizeThreshold(CourseAssessment assessment, BigDecimal raw) {
        if (raw == null) {
            return null;
        }
        if (usesBandScale(assessment)) {
            return normalizeBand(raw);
        }
        return raw.setScale(1, RoundingMode.HALF_UP);
    }

    public static BigDecimal clampBandScore(BigDecimal value, CourseAssessment assessment) {
        if (value == null) {
            return null;
        }
        BigDecimal cap = resolveScoreCap(assessment);
        if (cap != null && value.compareTo(cap) > 0) {
            return cap;
        }
        if (usesBandScale(assessment)) {
            return normalizeBand(value);
        }
        return value;
    }

    public static BigDecimal resolveDefaultMaxScore(
            AssessmentType type,
            AssessmentSkill skill,
            AiEvaluationMode evaluationMode
    ) {
        if (usesBandScale(type, skill, evaluationMode)) {
            return MAX;
        }
        return BigDecimal.TEN;
    }

    public static BigDecimal normalizeConfiguredMaxScore(
            BigDecimal maxScore,
            AssessmentType type,
            AssessmentSkill skill,
            AiEvaluationMode evaluationMode
    ) {
        if (maxScore == null) {
            return resolveDefaultMaxScore(type, skill, evaluationMode);
        }
        if (!usesBandScale(type, skill, evaluationMode)) {
            return maxScore;
        }
        if (maxScore.compareTo(MAX) > 0) {
            return MAX;
        }
        return maxScore;
    }

    public static BigDecimal normalizeConfiguredPassingScore(
            BigDecimal passingScore,
            AssessmentType type,
            AssessmentSkill skill,
            AiEvaluationMode evaluationMode
    ) {
        if (passingScore == null) {
            return null;
        }
        if (!usesBandScale(type, skill, evaluationMode)) {
            return passingScore;
        }
        return normalizeBand(passingScore);
    }
}
